### PR TITLE
[KYUUBI #2240][SUB-TASK] Skip add metadata manager if frontend does not support rest

### PR DESCRIPTION
**Pull request by ulysses-you**
### _Why are the changes needed?_

metadata manager is only used for batch session for rest frontend.

metadata manager is a heavy class, it will initialize metadata table and some of thread pool.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 1]
```diff
@@ -165,6 +173,11 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
     getSessionOption(sessionHandle).map(_.asInstanceOf[KyuubiBatchSessionImpl]).orNull
   }
 
+  private def metadataManager: MetadataManager = {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:176]: we can use metadataManager.foreach for all operations

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala]: changed

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 2]
```diff
@@ -45,14 +45,22 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
   // this lazy is must be specified since the conf is null when the class initialization
   lazy val sessionConfAdvisor: SessionConfAdvisor = PluginLoader.loadSessionConfAdvisor(conf)
   val applicationManager = new KyuubiApplicationManager()
-  private val metadataManager = new MetadataManager()
+  private var metadataManagerOpt: Option[MetadataManager] = None
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:48]:
```suggestion
  private lazy val metadataManager: Option[MetadataManager] = {
    // metadata manager is only used for batch session so if frontend does not support rest
    // it's unused.
    if (conf.get(FRONTEND_PROTOCOLS).map(FrontendProtocols.withName)
      .contains(FrontendProtocols.REST)) {
      Some(new MetadataManager())
    } else {
      None
    }
  }
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 3]
```diff
@@ -45,14 +45,22 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
...
-    addService(metadataManager)
+
+    // metadata manager is only used for batch session so if frontend does not support rest
+    // it's unused.
+    if (conf.get(FRONTEND_PROTOCOLS).map(FrontendProtocols.withName)
+        .contains(FrontendProtocols.REST)) {
+      metadataManagerOpt = Some(new MetadataManager())
+      addService(metadataManagerOpt.get)
+    }
+
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:63]:
```suggestion
   metadataManager.foreach(addService)
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 4]
```diff
@@ -166,23 +174,33 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
   }
 
   def insertMetadata(metadata: Metadata): Unit = {
-    metadataManager.insertMetadata(metadata)
+    metadataManagerOpt.foreach { metadataManager =>
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:177]: metadataManager.foreach(_.insertMetadata(metadata))

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 5]
```diff
@@ -45,14 +45,22 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
...
+  private var metadataManagerOpt: Option[MetadataManager] = None
 
   private var limiter: Option[SessionLimiter] = None
 
   override def initialize(conf: KyuubiConf): Unit = {
     addService(applicationManager)
     addService(credentialsManager)
-    addService(metadataManager)
+
+    // metadata manager is only used for batch session so if frontend does not support rest
```
[the remark this answers is not in the document]
Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala]: addressed comments

**Comment by turboFei** [comment]
```
FlinkOperationSuite:
org.apache.kyuubi.it.flink.operation.FlinkOperationSuite *** ABORTED ***
  java.lang.NullPointerException:
  at org.apache.kyuubi.session.KyuubiSessionManager.metadataManager$lzycompute(KyuubiSessionManager.scala:51)
```

**Comment by ulysses-you** [comment]
it seems we can not use kyuubi conf until `super.initialize`

**Comment by yaooqinn** [comment]
this.conf = conf

**Comment by ulysses-you** [comment]
test passed ! cc @turboFei @yaooqinn

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 6]
```diff
@@ -45,14 +45,25 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
...
+    }
+  }
 
   private var limiter: Option[SessionLimiter] = None
 
   override def initialize(conf: KyuubiConf): Unit = {
     addService(applicationManager)
     addService(credentialsManager)
-    addService(metadataManager)
+    this.conf = conf
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:64]: nit: to L62

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 7]
```diff
@@ -45,14 +45,25 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
...
+    metadataManager.foreach(addService)
+
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:66]: nit: remove blank line

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 8]
```diff
@@ -193,16 +204,18 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
       endTime: Long,
       from: Int,
       size: Int): Seq[Batch] = {
-    metadataManager.getBatches(batchType, batchUser, batchState, createTime, endTime, from, size)
+    metadataManager.map(
+      _.getBatches(batchType, batchUser, batchState, createTime, endTime, from, size))
+      .orNull
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:209]: is orNull a good choice, where NPE lurks

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:208]: I just follow the semantics of original code, the test will fail if change to Seq.empty.. I do not want to do more change in this pr so use `orNull`.

cc @turboFei

**Comment by ulysses-you** [comment]
@yaooqinn @turboFei any concern for this pr ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala (hunk 8 above)
Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:208]: which is the failed test

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:208]: Not sure which code cause test failed .. I wil try to fix it in this pr

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala]: @yaooqinn  it's ok now

**Review by yaooqinn — APPROVED** [review]
LGTM

**Comment by ulysses-you** [comment]
thank you for review, merging to master

**Pull request closed without merging**